GPT/beta-commands/gpt-function-calling.py

```diff
@@ -0,0 +1,165 @@
+                                "description": "The text to display",
+                            }
+                        },
+                        "required": ["str"],
+                    },
+                },
+            },
+            {
+                "type": "function",
+                "function": {
+                    "name": "search_for_command",
+                    "description": "search_for_command(str: string) - this searches for a command in the VSCode command palette. If I ask you to do something, please use this command to search for an appropriate command.",
+                    "parameters": {
+                        "type": "object",
+                        "properties": {
+                            "str": {
+                                "type": "string",
+                                "description": "The command to search for",
+                            }
+                        },
+                        "required": ["str"],
+                    },
+                },
+            }
+        ],
+        "max_tokens": 2024,
+        "temperature": 0.6,
+        "n": 1,
+        "stop": None,
+        "model": settings.get("user.openai_model"),
+    }
+
+
+    response = requests.post(url, headers=headers, data=json.dumps(data))
+
+    if response.status_code == 200:
+        notify("GPT Task Completed")
+        message = response.json()["choices"][0]["message"]
+        process_function_calls(insert_response, message)
+
+        content = (message["content"] or "").strip()
+        if len(content) != 0:
+            display_response(content)
+
+    else:
+        notify("GPT Failure: Check API Key, Model, or Prompt")
+        raise Exception(f"GPT Failure at POST request: {response.json()}")
+
+def process_function_calls(insert_response, message):
+    try:
+        tool_calls = message["tool_calls"]
+    except Exception as e:
+        notify(f"No tool calls were found in LLM response")
+        return
+
+    for tool in tool_calls:
+
+        try: 
+            first_argument = tool['function']['arguments']
+            first_argument = json.loads(first_argument)['str']
+        except Exception as e:
+            notify(f"Argument JSON was malformed: {e}")
+            break
+
+        match tool['function']['name']:
+            case 'insert':
+                insert_response(first_argument)
+            case 'display':
+                display_response(first_argument)
+            case 'notify':
+                notify_user(first_argument)
+            case 'search_for_command':
+                search_for_command(first_argument)
+
+
+@mod.action_class
+class UserActions:
+    def gpt_can_you(utterance: str, selected_text: str) -> str:
+        """Run a query with function calls and insert the result"""
+        return gpt_function_query(utterance, selected_text, actions.user.paste)
+
+    def gpt_can_you_cursorless(utterance: str, text_to_process: str, cursorless_destination: any):
```

Comment:
note that `cursorless_destination` might be `0` now so we need to check for that case

Reply:
Still need to fix this. Not exactly sure what we want to happen if the destination is invalid. I am not super clear on the usage.

Reply:
Handled the 0 case.